\documentclass{bioinfo}
\copyrightyear{2012}
\pubyear{2012}

\usepackage{subfig}
\usepackage{listings}
\usepackage{ifthen}

\lstset{language=Java,
morendkeywords={String, Throwable}
captionpos=b,
basicstyle=\scriptsize\ttfamily,%\bfseries
stringstyle=\color{darkred}\scriptsize\ttfamily,
keywordstyle=\color{royalblue}\bfseries\ttfamily,
ndkeywordstyle=\color{forrestgreen},
numbers=left,
numberstyle=\scriptsize,
% backgroundcolor=\color{lightgray},
breaklines=true,
tabsize=2,
frame=single,
breakatwhitespace=true,
identifierstyle=\color{black},
% morecomment=[l][\color{forrestgreen}]{//},
% morecomment=[s][\color{lightblue}]{/**}{*/},
% morecomment=[s][\color{forrestgreen}]{/*}{*/},
commentstyle=\ttfamily\itshape\color{forrestgreen}
% framexleftmargin=5mm,
% rulesepcolor=\color{lightgray}
% frameround=ttff
}

% MACROS:

\newcommand{\TODO}[1]{\textcolor{red}{\textbf{#1}}}
\newcommand{\AbstractDESSolver}{\texttt{Abstract\-DES\-Solver}}
\newcommand{\SBMLinterpreter}{\texttt{SBML\-interpreter}}

\hyphenation{
  % TODO hypens for regular words
}

% some nice colors
\definecolor{royalblue}{cmyk}{.93, .79, 0, 0}
\definecolor{lightblue}{cmyk}{.10, .017, 0, 0}
\definecolor{forrestgreen}{cmyk}{.76, 0, .76, .45}
\definecolor{darkred}{rgb}{.7,0,0}
\definecolor{winered}{cmyk}{0,1,0.331,0.502}
\definecolor{lightgray}{gray}{0.97}

\begin{document}
\firstpage{1}

\title[SBML Simulation Core Library]{SBML Simulation Core Library: the Java
library for numerical computation with SBML} \author[Dr\"ager
\textit{et~al.}]{%
Andreas Dr\"ager\,$^{1,*}$, Roland Keller\,$^{1}$, 
Alexander D\"orr\,$^{1}$,
Akito Tabira\,$^{2}$,
Akira Funahashi\,$^{2}$,
\TODO{Nicolas Rodriguez\,$^{3}$,
Nicolas Le Nov\`{e}re\,$^{3}$} and
Andreas Zell\,$^1$\footnote{to whom correspondence should be addressed}}
\address{$^{1}$Center for Bioinformatics Tuebingen (ZBIT), University of
Tuebingen, T\"ubingen, Germany\\
$^{2}$Keio University, Graduate School of Science and Technology, Yokohama,
Japan\\
\TODO{$^{3}$European Bioinformatics Institute, Wellcome Trust Genome Campus,
Hinxton, Cambridge, UK\\}}

\history{Received on XXXXX; revised on XXXXX; accepted on XXXXX}

\editor{Associate Editor: XXXXXXX}

\maketitle

\begin{abstract}

\section{Motivation:}
The Systems Biology Markup Language (SBML) has been designed as a medium for the
exchange of quantitative models of biological networks. To this end, SBML
defines a rich set of methods to specify differential equation systems for
these networks. As an XML-based standard, however, models encoded in SBML
require the availability of suitable software implementations in order to
perform computer simulations and further analyses. Many solvers for SBML have
been implemented and are available, but the vast majority of these is either
part of a larger software suite and can therefore not be easily integrated into
custom programs, or implemented in programing languages, which are either
platform-dependent or even require a commercial license for their
execution.

\section{Results:}
We here present the SBML Simulation Core library, an exhaustive implementation
of the SBML standard for numerical calculation and network simulation in form of
an application programming library for the Java\texttrademark{} programming
language. This library has been implemented as a community-driven project. Its
designed has the potential to be extended for the simulation of further
standard network formats in systems biology. It does not depend on any
commercial library and is entirely written in Java without the necessity to
include any platform-dependent wrappers or libraries. To demonstrate its
capabilities, it has been benchmarked against the entire SBML Test Suite and
also been used to simulate all models of the BioModels.net database.

\section{Availability:}
Source code, binaries, and documentation for SBML Simulation Core can
be freely obtained under the terms of the LGPL 3.1 from the website
\href{http://www.cogsys.cs.uni-tuebingen.de/software/SBMLsimulator}{http://www.cogsys.cs.uni-tuebingen.de/software/SBMLsimulator}.

\section{Contact:}
\TODO{We need a mailing list for this}
\href{mailto:andreas.draeger@uni-tuebingen.de}{andreas.draeger@uni-tuebingen.de}

\section{Supplementary information:}
% TODO: Provide additional material
Supplementary data are available at Bioinformatics online.

\end{abstract}

\section{Introduction}

The modeling language SBML (Systems Biology Markup Language,
\citealt{Hucka2003}) constitutes an important \emph{de facto} standard for the
exchange of biochemical network models.
SBML defines a set of data structures and provides rules about how to interpret
and simulate these kinds of models.

A model stored in SBML may combine an ordinary differential equation system,
which is the basis for numerical simulation, with additional elements such as
rules and events.  These elements further influence the system. For instance,
an event takes place if a certain trigger condition becomes true. Whenever this
happens, event assignments may change the values of model components, such as
parameter values or compartment sizes. Rules can directly assign new values to
their objectives, e.g., the concentration of a reacting species.

In many cases, model parameters with uncertain values, such as kinetic
constants, need to be estimated using heuristic optimization procedures in
order to minimize the distance between model output and experimental
measurement data.

For model building, simulation, and parameter calibration processes, dedicated
software implementations are indispensable. There are several larger tools with
a graphical user interface written in Java\texttrademark{}, C or MATLAB that
fulfil those important requests:
\citep{Loew2001}, \citep{Myers2009}, \citep{Maiwald2008}, \citep{Hoops2006}, 
\citep{SBT_Schmidt2006}, \citep{Bergmann06}.
But it is difficult to integrate the contained simulation routines in other programs.
The SBML ODE Solver Library \citep{Machne2006}, which is written in C and based
on the libSBML library \citep{Bornstein2008},  provides such a simulation
routine. The SUNDIALS package is used for solving the differential equation system given in the model.

In contrast to that we here present a platform-independent simulation API in
Java that interprets the content of such a model and predicts the dynamic behavior of the model’s
components. It is the first simulation routine which is based on the
Java\texttrademark{} library JSBML \citep{Draeger2011b}, a specifically
developed data structure to read and write models from and into SBML files and
to deal with their structure in memory. The API contains an interpreter for the
SBML model and several different ODE solvers. The generic library is completely
decoupled from any graphical user interface and can hence easily be integrated as an API (Application Programming Interface) into third-party programs under the terms of the LGPL Version 3.
\TODO{Describe this as a potential application.}
The SBML simulation core library has already been integrated into the widely
used program CellDesigner version 4.2 \citep{Funahashi2003} as an internal simulation
library.
%Secondly, a graphical and command-line user interface that provides
%a connection to the heuristic optimization framework EvA2 \citep{Kron10EvA2}.
% The combination of SBMLsimulator and EvA2 \citep{Kron10EvA2} estimates the values of all parameters with
%respect to given time-series of metabolite or gene expression values. 

\begin{methods}
\section{Implementation}
In this API a large collection of differential equation solvers and an
interpreter for SBML are provided. The solvers and the interpretation of SBML
are independent of each other.

All the solver classes are derived from the abstract class \AbstractDESSolver.
The Eulerian and the Runge-Kutta solver are very fast, but not suitable
for some differential equation systems. Several solvers of the Apache Commons
Math library are integrated with the help of wrapper classes. The Rosenbrock
solver \citep{Kotcon2011} is slower than the other solvers, but it can be used
for solving stiff differential equation systems.

The class \SBMLinterpreter{} stores and processes a given SBML model. For a given
value of the ODE system and a given time it returns the current set of
derivatives of the variables. 

An extremely important aspect in the interpretation of SBML models is the
determination of the exact time at which an event occurs, as this can have a
high influence on the precision of the values of the system variables. We have
therefore adapted the Rosenbrock solver, which is an integrator with an adaptive
step size, to a very precise timing of the events.

The API contains an efficient interpreter for MathML expressions that are
contained in kinetic laws, rules and events. The nodes of the syntax tree of
those expressions depend on the current simulation time and the given values of
the ODE system. If the time or any of these values has changed, the value of
the node has to be recalculated. At the beginning of the simulation the syntax
trees of all kinetic laws, rules and events are restructured and merged to one
large tree that contains equivalent nodes only once. This leads to a decreasing
computation time during the simulation.
\end{methods}

\section{Results}
Our implementation has been successfully tested with the SBML test suite (see
\href{http://sbml.org/Software/SBML_Test_Suite}{http://sbml.org/Software/SBML\_Test\_Suite})
and the models of the biomodels.net database: With the Rosenbrock solver all
models of the test suite are simulated correctly and all but three models of the
\href{http://biomodels.net}{BioModels.net} database \citep{Novere2006a} can be
simulated in acceptable time.

%\begin{table}[!t]
%\processtable{This is table caption\label{Tab:01}}
%{\begin{tabular}{llll}\toprule
%head1 & head2 & head3 & head4\\\midrule
%row1 & row1 & row1 & row1\\
%row4 & row4 & row4 & row4\\\botrule
%\end{tabular}}{This is a footnote}
%\end{table}



%\begin{figure}[!tpb]%figure1
%%\centerline{\includegraphics{fig01.eps}}
%\caption{Caption, caption.}\label{fig:01}
%\end{figure}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%
%     please remove the " % " symbol from \centerline{\includegraphics{fig01.eps}}
%     as it may ignore the figures.
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\section{Conclusion}
The SBMLsimulator API is an efficient Java tool for the simulation of
differential equation systems given in SBML. It can be easily integrated in
larger applications, which display the simulation results or optimize some
parameters of ODE models. 
With the given program structure the support of other model formats (e.g.,
CellML, \citealt{Lloyd2004}) can be added by only implementing an interpreter
for each format.


\section*{Acknowledgement}

The authors are grateful to Beky Kotcon, Samantha Mesuro, Daniel Rozenfeld, Anak
Yodpinyanee, Andres Perez, Eric Doi, Richard Mehlinger, Steven Ehrlich, Martin
Hunt, George Tucker, Peter Scherpelz, Aaron Becker, Eric Harley, and Chris
Moore, Harvey Mudd College, Claremont, California, USA, for providing an
Java implementation of Rosenbrock's numerical ODE solver. 

\paragraph{Funding\textcolon} 
The Federal Ministry of Education and Research (BMBF, Germany) in the project
Virtual Liver Network (grant number 0315756).

\paragraph{Conflict of Interest\textcolon} none declared.

\bibliographystyle{natbib}
%\bibliographystyle{achemnat}
%\bibliographystyle{plainnat}
%\bibliographystyle{abbrv}
%\bibliographystyle{bioinformatics}
%
%\bibliographystyle{plain}
%
\bibliography{document}

\end{document}
